### ISSUE TITLE
[question] How web friendly is WinJS

### ISSUE THREAD

**Issue by SteveALee** [winjs/winjs#1551]
I'd really like to use the WinJS in my latest web project/product and have been tracking it since the first release and Kraig's ebook. I love the multi input modes and quality look and feel. It's even better to see you explicitly support some accessibility now.

However I have a nagging question about how web friendly it is. By that I mean

* Does it support a wide range of devices and browsers - including older ones?
* How is it on poor connectivity as often experienced in mobile?

Or is it mainly used in offline, file based scenarios like WIndows JS apps?

There's obviously a lot of issues wrapped up in each of those but my main concerns are

## Browser support
The FAQ bravely claims it will work in **any** browser but the BrowserSupport page is more realistic and list latest versions of the evergreen browsers. That page also claims it may work on others. Do you have any more experience you can relate. What if someone is a cruddy old using an old phone in Africa? What is performance like on older browser (especially Android)

## Poor connectivity
Many javascript frameworks suffer from being unusable until all the code has loaded, parsed and run. Does WinJS improve on this in any way? I guess we can make custom builds to remove parts we don't use but thats no substitute for a progressive implementation.

I guess what I'm really thinking of is Progressive Enhancement. To me that doesn't just mean it will workout with Javascript (not so feasible). Rather, I want the user experience to depend on the device capabilities (responsive design, wide browser support) , the connection reliability (offline etc) and the user preferences and context (accessibility). Ideally the user wil always get somehting they can use, within some reasonable limits of course.

I want to provide a minimal experience if the context requires it, perhaps without WInJS at all if it doesn't fall back itself. But how could I determine the context can support WInJS and how much work will that be?

JQueryMobile is the only framework I found that explicitly supports PE but I wonder how usable plain <a> links are on a small touch screen. They do have a useful Graded Browser Support system but I bet that is plenty of work to maintain,

I recall reading in old WinJS issue a year or so back that no one was using WInJS in a Web contex. Is that still true? I see you now have the web framework adapters which hints it is being used in wider context now,

Any thoughts or comments much appreciated. Thanks.

PS the BrowserSupport page mentions a status page in the last paragraph but I cannot find it. I did find the travis output on the old demo pages, though interestingly they mention IE on Windows 8.1 and not Edge on 10.

**Comment by toddpi314** [comment]
@SteveALee I can tell you first hand that I have personally used WinJS for web-based applications. In most cases, the app is hosted remotely with WinJS as the "UI Framework" and then mounted to a device context (Connected Devices, Mobile Web View or Browser)

I would not recommend pre-v4.0 for these efforts, as many of the important non-web-centric focus features were lacking (or buggy).

In terms of framework perf, WinJS is just UI and is pretty light considering what it does. You have to bring your own patterns (PE, External View Fragments, etc). Most is front-loaded on script reference, so you won't get away with shipping without serious minification/closure'ing, but after that its pretty snappy. Load that bad boy up in webpack with uglify and google closure and you are off to the races without many issues.

If you end up using the built in Navigation framework, you do have a pretty flexible option to resolve views ad-hoc or baked, which is helpful in cases that you are using WebPack and don't what the view templates being loaded too early. I have even seen teams render out the WinJS Views on a server and have the client just navigate around a series of remote links. Its all good.

**Gotchas**:
- Focus is bleeding edge and there are still some cases to wrestle (imported from Xbox One UI framework and working pretty well)
- UI uses virtualization that sometimes can render controls (ListView, Pivot) with different behaviors on all of the insane device resolutions out in the wild. Make sure you test your responsive on the target form factors. Things like "Item Template Clipping" and "Incremental scrolling" can occur and be a bit difficult to tackle.
- WinJS is just a UI framework. Expect to bring your own architecture (Angular, React, Heck, i wrote my own MVVM https://github.com/DeepElement/winjsrocks)

In terms of browser support, we were able to run v4.* on webkit from back in 2009 with about 15 hours of effort to extend styles. Backwards compatibility isn't that bad with the expectation of extending the base styles. I believe we should thank the TypeScript strategy for that.

**Biggest Gotcha of them all**:
 MS team maintaining project clearly favor changes benefiting legacy windows developers over web users
With thousands of apps using older WinJS (Windows 8, Windows 8.1, XboxOne current gen), the priorities tend to favor what those users request more than the few ambitious web adopters. 

Hope that helps.
Todd

**Comment by toddpi314** [comment]
@SteveALee Something also that should be called out...

The Keyboard/Accessibility story for WinJS is just awesome. 

The entire control suite is navigable via Keyboard/Gamepad/Mouse/Touch on all platforms. By that, I don't mean they support "TabIndex"; I mean they wrote a topological mapping algorithm that plots intuitive visual focus transitions based on the DOM state. 

I have personally written this type of logic on other platforms and it gets very complicated with your View isn't a simple state diagram. 

At least give it a shot and test out the focus features alone, if nothing else.

**Comment by toddpi314** [comment]
Citing a real world example:

We just had a bug come in that focus doesn't always give access to the last item template in a list view on Chrome 34.0.0.0 (Android 4.2.2). Not reproducible on Blink Engine latest (Android 5.1).

We see small variations like this trickle in purely related to the focus logic re-adjusting scroll position of containers based on the focused item's visible rect.

**Comment by SteveALee** [comment]
Thanks very Much @toddpi314. That **is** helpful. A few comments:

> In most cases, the app is hosted remotely with WinJS as the "UI Framework" and then mounted to a device context (Connected Devices, Mobile Web View or Browser)

So you deploy as web app or Win App? DO you have a link? 

Do you do any explicit local caching, or are all asset fetches remote round trips (possibly served from automatic caches)?

> I would not recommend pre-v4.0 for these efforts, as many of the important non-web-centric focus features were lacking (or buggy).

Yeah, I gave up a while back. Actually twice as first attempt was Win JS app in early days of 8.1 when dev docs were very patchy and sandboxing got confusing.

> Most is front-loaded on script reference, so you won't get away with shipping without serious minification/closure'ing, but after that its pretty snappy. Load that bad boy up in webpack with uglify and google closure and you are off to the races without many issues.

Ok so not idea on bad connectivity. Doesn't look like the custom build lets you strip out stuff (at least from wiki docs). Using webpack uglify for production but hadn't thought of google closure so thanks for tip.

> Focus is bleeding edge and there are still some cases to wrestle (imported from Xbox One UI framework and working pretty well).

Not surprising as browser diffs and probably not user tested as much. I'm a heavy keyboard user so tend to notice.

> WinJS is just a UI framework. Expect to bring your own architecture (Angular, React, Heck, i wrote my own MVVM https://github.com/DeepElement/winjsrocks)

I was looking at riotjs as like small lib approach. however it's server-side rendering strips out events so no good for PE. Must revisit react.

> MS team maintaining project clearly favor changes benefiting legacy windows developers over web users. With thousands of apps using older WinJS (Windows 8, Windows 8.1, WinPhone, XboxOne current gen), the priorities tend to favor what those users request more than the few ambitious web adopters. 

Again not surprising. However the fact that recent Micorsoft public events have almost completely neglected JS App dev makes we worry it might get even worse.

> The Keyboard/Accessibility story for WinJS is just awesome. 
> The entire control suite is navigable via Keyboard/Gamepad/Mouse/Touch on all platforms. By that, I don't mean they support "TabIndex"; I mean they wrote a topological mapping algorithm that plots intuitive visual focus transitions based on the DOM state. 

That is a key reason I'm interested so it's very good to hear your confirmation! Given that it's surprising the keyboard accessibility of windows 10 and MS store apps have gapping holes. I guess they don't use WinJS - hehe.

thanks again.

So I'm left wondering about support on older browser and less capable devices.

**Comment by SteveALee** [comment]
@toddpi314

> In terms of browser support, we were able to run v4.* on webkit from back in 2009 with about 15 hours of effort to extend styles. Backwards compatibility isn't that bad with the expectation of extending the base styles. I believe we should thank the TypeScript strategy for that.

I skipped this earlier. That is encouraging. Assuming you meant '[forward compatibility](https://en.wikipedia.org/wiki/Forward_compatibility)'? ;) Confusing terms these. 

By 'TypeScript' strategy do you mean transpiling to ES5 or something else?

**Comment by SteveALee** [comment]
So @scottjehl sums up exactly [what I mean by 'web friendly'](https://www.filamentgroup.com/lab/delivering-responsibly.html)

**Comment by SamuelEnglard** [comment]
I'm using WinJS 4.0.1 on my blog (blog.englard.net). I primarily chose it for the Accessibility features and the visual look it gives.

I personally made the choice to accept that older browsers would not be able to see my site, though I do provide a fallback to show them a message. I'm currently working though on a better solution to give them kind of a "WinJS Light" version.

What I have learnt is that most of the "modern" stuff that WinJS requires is CSS3 stuff. the JavaScript and HTML is very old browser friendly.

In theory the HTML for basic UI should look ok without WinJS running, so the site will still be useable in some sense, though it will be limited. If a site is mostly content reading based then WinJS being loaded in the background slowly and then applied won't be a problem. For more app like websites it might be an issue I think.

**Comment by SteveALee** [comment]
@SamuelEnglard Thanks for a very helpful answer

> I primarily chose it for the Accessibility features and the visual look it gives.
:1: and the clear approach to the UI appears to have some cog a11y advantages too.

> I'm currently working though on a better solution to give them kind of a "WinJS Light" version.

That' my sticking point. I want plain HTML,CSS, links and forms to work to some extent as a fallback

> What I have learnt is that most of the "modern" stuff that WinJS requires is CSS3 stuff. the JavaScript and HTML is very old browser friendly.

That is useful information. And long may it remain so ;)
I did look for CSS3 shims an polyfils that are also mobile friendly but had little luck.

> In theory the HTML for basic UI should look ok without WinJS running

I had assumed that wouldn't as it appears to be so reliant on JS. Perhaps I should actually try.

For now I'm starting with jQueryMobile with a plan to enhance with WinJS at some point.

**Comment by SamuelEnglard** [comment]
@SteveALee Happy to help

> Your blog looks great.

Thanks.

> That' my sticking point. I want plain HTML,CSS, links and forms to work to some extent as a fallback

What I'm thinking may work (but I need to look more into) is if some WinJS work could be done server side, so that on page load the page has more structure too it.

> I did look for CSS3 shims an polyfils that are also mobile friendly but had little luck.

I actually tried to see how the site would render if I just forced it to work in older browsers without CSS3. It actually worked pretty well. Main reason I ended up not doing it was that I myself needed CSS3 to make the site use work the way I wanted. Being a personal blog I felt that I could take the risk of cutting off people on older browsers.

> For now I'm starting with jQueryMobile with a plan to enhance with WinJS at some point.

I have a personal dislike of anything jQuery but I fully realize that many of my dislikes are subjective and/or personal.
